docs/design/ROW_FORMAT_V2.md

```diff
@@ -0,0 +1,78 @@
+
+# Proposal: A new storage row format for efficient decoding
+
+- Author(s):     Yuxing ZHOU
+- Last updated:  2018-07-17
+
+## Abstract
+
+This proposal proposes a new row format for efficient decoding.
+
+## Background
+
+In the current row format, each column is encoded with the column ID followed by the column value and multiple columns are concatenated together. In order to find a specific column in the row, we need to start from the beginning and decode each column ID and the length of each value until we find the matched column. This decoding method is inefficient especially when the number of columns in a row is not small and the complexity is O(N).
+
+## Proposal
+
+The new format is defined as follows:
+
+* 1 byte codec version. Its value starts from 128, so we can upgrade TiDB in a compatible way.
+
+* 1 byte row type flag containing the following attribute:
+
+	- 1 bit IsBig
+	A row is big if the max column ID is greater than 255 or the total size of the values is greater than 65535. We can add more types in this byte later if needed, for example storing 2 bytes column ID.
+
+
+* 2 bytes number of not-null columns
+
+* 2 bytes number of null columns
+
+We can store 65536 columns in a row and this number is much larger than the upper limit in TiDB.
+
+* not null column ID array
+If there are too many columns, each column ID takes 4 bytes, otherwise the column ID takes 1 byte. The IDs array is sorted in an ascending order.
```

Comment:
Please add a line before this paragraph and this paragraph should be indented with 4 spaces.